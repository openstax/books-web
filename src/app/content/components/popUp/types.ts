import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
