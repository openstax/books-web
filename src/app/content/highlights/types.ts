import { Highlight, HighlightsSummary } from '@openstax/highlighter/dist/api';

export type HighlightData = Highlight;

export type CountsPerSource = Exclude<HighlightsSummary['countsPerSource'], undefined>;

export interface State {
  myHighlightsOpen: boolean;
  enabled: boolean;
  focused?: string;
  highlights: null | HighlightData[];
  summary: {
    filters: {
      colors: string[];
      chapters: string[];
    },
    pagination: {
      // even though we're manually splitting our requests out into smaller batches of sources,
      // if a source has more than [numPage] highlights the api will still paginate. in order to
      // access the subsequent pages, we need to store the original sources here, and continue to
      // use them until we run out of pages, then grab the next batch of sources. on the LAST
      // page the response size may be smaller than expected, in this case the next batch of
      // sources should be immediately queried to fill the remaining [numPage]. use
      // paginationUtils.getNextPageSources to get the next batch of sources
      sources: string[];
      page: number;
    } | null,
    loading: boolean;
    // totalCounts should reflect the UNFILTERED state of user's data, used in representing
    // available filter options. this should be updated on create/delete and after auth change
    totalCounts: {[key: string]: number};
    // filteredTotalCounts should reflect the FILTERED state of user's data, used in calculating
    // pagination requests. this should be updated after filter change before requesting
    // filtered highlights (because you need this to query them)
    filteredTotalCounts: {[key: string]: number};
    highlights: {
      [chapterId: string]: {[pageId: string]: HighlightData[]}
    };
  };
}
