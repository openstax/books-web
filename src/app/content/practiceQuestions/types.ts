import { LinkedArchiveTree, LinkedArchiveTreeSection } from '../types';

export interface State {
  isEnabled: boolean;
  summary: PracticeQuestionsSummary | null;
  open: boolean;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export type PracticeQuestionsLocationFilters = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;
