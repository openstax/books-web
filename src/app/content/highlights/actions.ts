import { createStandardAction } from 'typesafe-actions';

export const enableHighlighting = createStandardAction('FeatureFlag/enableHighlighting')<void>();
