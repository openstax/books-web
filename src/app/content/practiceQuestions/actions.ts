import { createStandardAction } from 'typesafe-actions';
import { PracticeQuestionsSummary } from './types';

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();
