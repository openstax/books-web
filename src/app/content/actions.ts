import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as searchActions from './search/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();

export const closeCallToActionPopup = createStandardAction('Content/closeCallToActionPopup')();
export const openCallToActionPopup = createStandardAction('Content/openCallToActionPopup')();
