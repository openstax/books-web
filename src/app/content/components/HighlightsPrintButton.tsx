import React from 'react';
import styled from 'styled-components/macro';
import { assertWindow } from '../../utils';
import { printSummaryHighlights } from '../highlights/actions';
import PrintButton from './Toolbar/PrintButton';

interface Props {
  className?: string;
  isLoading: boolean;
  shouldFetchMore: boolean;
  loadHighlightsAndPrint: typeof printSummaryHighlights;
}

// tslint:disable-next-line:variable-name
const StyledPrintButton = styled(PrintButton)`
  min-width: auto;
  height: max-content;
  margin-left: auto;
  cursor: ${({isLoading}) => isLoading ? 'wait' : 'pointer'};
`;

// tslint:disable-next-line:variable-name
export default ({className, isLoading, loadHighlightsAndPrint, shouldFetchMore}: Props) => {
  return <StyledPrintButton
    className={className}
    isLoading={isLoading}
    onClick={() => shouldFetchMore ? loadHighlightsAndPrint() : assertWindow().print()}
    data-testid='hl-print-button'
    disabled={isLoading}
  />;
};
