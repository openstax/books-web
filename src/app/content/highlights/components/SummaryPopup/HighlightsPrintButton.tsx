import flow from 'lodash/fp/flow';
import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components/macro';
import { AppState, Dispatch } from '../../../../types';
import PrintButton from '../../../components/Toolbar/PrintButton';
import { printSummaryHighlights } from '../../actions';
import { hasMoreResults } from '../../selectors';

interface Props {
  className: string;
  shouldFetchMore: boolean;
  print: typeof printSummaryHighlights;
}

// tslint:disable-next-line:variable-name
const HighlightsPrintButton = ({className, print, shouldFetchMore}: Props) => {
  return <PrintButton
    className={className}
    onClick={() => print({shouldFetchMore})}
    data-testid='hl-print-button'
  />;
};

// tslint:disable-next-line:variable-name
const ConnectedHighlightsPrintButton = connect(
  (state: AppState) => ({
    shouldFetchMore: hasMoreResults(state),
  }),
  (dispatch: Dispatch) => ({
    print: flow(printSummaryHighlights, dispatch),
  })
)(HighlightsPrintButton);

export default styled(ConnectedHighlightsPrintButton)`
  min-width: auto;
  height: max-content;
  margin-left: auto;
`;
