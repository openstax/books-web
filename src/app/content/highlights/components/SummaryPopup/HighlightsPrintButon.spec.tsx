import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../../test/createTestStore';
import MessageProvider from '../../../../MessageProvider';
import { Store } from '../../../../types';
import { printSummaryHighlights, receiveSummaryHighlights } from '../../actions';
import HighlightsPrintButton from './HighlightsPrintButton';

describe('HighlightsPrintButton', () => {
  let store: Store;
  let storeDispatch: jest.SpyInstance;

  beforeEach(() => {
    store = createTestStore();
    storeDispatch = jest.spyOn(store, 'dispatch');
  });

  it('matches snapshot', () => {
    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <HighlightsPrintButton />
      </MessageProvider>
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('dispatches action to print highlights', () => {
    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <HighlightsPrintButton />
      </MessageProvider>
    </Provider>);

    renderer.act(() => {
      store.dispatch(receiveSummaryHighlights({}, {} as any));
    });

    const renderedPrintButton = component.root.findByProps({'data-testid': 'hl-print-button'});

    renderer.act(() => {
      renderedPrintButton.props.onClick();
    });

    expect(storeDispatch).toHaveBeenCalledWith(printSummaryHighlights({shouldFetchMore: true}));
  });
});
