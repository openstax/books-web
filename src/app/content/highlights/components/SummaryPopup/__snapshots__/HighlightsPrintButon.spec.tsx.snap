// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightsPrintButton matches snapshot 1`] = `
.c3 {
  display: inline-block;
  vertical-align: -.125em;
  overflow: hidden;
}

.c1 {
  cursor: pointer;
  border: none;
  padding: 0;
  background: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #5e6062;
  height: 100%;
  min-width: 45px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
}

.c1:hover,
.c1:focus {
  color: #424242;
}

.c4 {
  font-weight: 600;
  font-family: Neue Helvetica W01;
  font-size: 1.6rem;
  line-height: 2.5rem;
  margin: 0 0 0 0.5rem;
}

.c2 {
  height: 2.5rem;
  width: 2.5rem;
  padding: 0.4rem;
}

.c0 {
  min-width: auto;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
  margin-left: auto;
}

@media screen and (max-width:75em) {
  .c4 {
    display: none;
  }
}

<button
  aria-label="print"
  className="c0 c1"
  data-testid="print"
  onClick={[Function]}
>
  <svg
    aria-hidden="true"
    className="c2 c3"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 512 512"
  >
    <path
      d="M448 192V77.25c0-8.49-3.37-16.62-9.37-22.63L393.37 9.37c-6-6-14.14-9.37-22.63-9.37H96C78.33 0 64 14.33 64 32v160c-35.35 0-64 28.65-64 64v112c0 8.84 7.16 16 16 16h48v96c0 17.67 14.33 32 32 32h320c17.67 0 32-14.33 32-32v-96h48c8.84 0 16-7.16 16-16V256c0-35.35-28.65-64-64-64zm-64 256H128v-96h256v96zm0-224H128V64h192v48c0 8.84 7.16 16 16 16h48v96zm48 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
      fill="currentColor"
    />
  </svg>
  <span
    className="c4"
  >
    Print
  </span>
</button>
`;
